src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -57,11 +97,82 @@ private void AddExceptionDiagnostic(Diagnostic diagnostic)
         /// </summary>
         public async Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync()
         {
-            // Invoke GetDiagnostics to populate the compilation's CompilationEvent queue.
-            ImmutableArray<Diagnostic> compilerDiagnostics = _compilation.GetDiagnostics(_cancellationToken);
+            ImmutableArray<Diagnostic> analyzerDiagnostics = await GetAnalyzerDiagnosticsAsync().ConfigureAwait(false);
+            return _compilation.GetDiagnostics(_cancellationToken).AddRange(analyzerDiagnostics).AddRange(_exceptionDiagnostics);
+        }
+
+        /// <summary>
+        /// Returns diagnostics produced by diagnostic analyzers from analyzing a single document.
+        /// <param name="model">Semantic model for the document.</param>
+        /// </summary>
+        public async Task<ImmutableArray<Diagnostic>> GetAnalyzerDiagnosticsFromDocumentAsync(SemanticModel model)
+        {
+            SyntaxTree documentTree = model.SyntaxTree;
+            Task<ImmutableArray<Diagnostic>> documentTask;
+
+            while (true)
+            {
+                Task latestAnalysisTask = _latestAnalysisTask;
+                if (latestAnalysisTask != null)
+                {
+                    await latestAnalysisTask.ConfigureAwait(false);
+                }
 
-            ImmutableArray<Diagnostic> analyzerDiagnostics = await _driver.GetDiagnosticsAsync().ConfigureAwait(false);
-            return compilerDiagnostics.AddRange(analyzerDiagnostics).AddRange(_exceptionDiagnostics);
+                lock (_analysisLock)
+                {
+                    if (!_documentTasks.TryGetValue(documentTree, out documentTask))
+                    {
+                        if (_completeAnalysisTask != null)
+                        {
+                            // Once complete analysis has begun, it is too late to start analysis of an individual document.
+                            return ImmutableArray<Diagnostic>.Empty;
+                        }
+
+                        if (latestAnalysisTask != _latestAnalysisTask)
+                        {
+                            // Another analysis task has started. Wait to start this one.
+                            // Analysis tasks are serialized because the driver can use all available cores
+                            // in one analysis and swamping the system with tasks is not advantageous.
+                            continue;
+                        }
+
+                        _analyzedSyntaxTrees = _analyzedSyntaxTrees.Add(documentTree);
```

Comment:
Suggestion (feel free to ignore):
Are we maintaining the `_analyzedSyntaxTrees` field only to track which analyzers have already executed syntax tree actions? If so, I think a simpler fix might be in the driver itself. We can have a `CompilationUnitStarted` event generated on a tree parse, and processing this event we invoke SyntaxTree action for the tree. Alternatively, we can just move the syntax tree action execution into the current CompilationUnitCompleted event processing.